flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MultiJoinTestPrograms.java

```diff
@@ -0,0 +1,461 @@
+                                            Row.ofKind(RowKind.INSERT, "3", "Brasília"))
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink")
+                                    .addSchema(
+                                            "user_id STRING",
+                                            "name STRING",
+                                            "order_id STRING",
+                                            "payment_id STRING",
+                                            "location STRING")
+                                    .addOption("changelog-mode", "I,UA,UB,D")
+                                    .consumedValues(
+                                            "+I[1, Gus, order0, payment1, London]",
+                                            "+I[1, Gus, order1, payment1, London]",
+                                            // NO I[2, Bob..] here because Bob was deleted
+                                            "+I[3, Nomad, order3, payment3, New York]",
+                                            // NO I[4, David..] here because David has only 5
+                                            // dollars, so  this is not true u.cash >= p.price
+                                            "+I[5, Eve, null, payment5, null]", // Here because
+                                            // payment price as
+                                            // -1
+                                            // NO I[7, Welcher..] here because Welcher has no
+                                            // payment
+                                            "+I[6, Frank, order6, payment6, null]",
+                                            // NO I[8, Welcher..] here because there is no order and
+                                            // no payment
+                                            "+I[9, Charlie Taylor, order9, payment9, Melbourne]", // Charlie Taylor because the name was updated
+                                            "+I[3, Nomad, order3, payment3, Brasília]" // New
+                                            // Location
+                                            // for user 3
+                                            )
+                                    .testMaterializedData()
+                                    .build())
+                    .runSql(
+                            "INSERT INTO sink "
+                                    + "SELECT u.user_id_0, u.name, o.order_id, p.payment_id, s.location "
+                                    + "FROM Users u "
+                                    + "LEFT JOIN Orders o ON u.user_id_0 = o.user_id_1 "
+                                    + "INNER JOIN Payments p ON u.user_id_0 = p.user_id_2 AND (u.cash >= p.price OR p.price < 0) "
+                                    + "LEFT JOIN Shipments s ON p.user_id_2 = s.user_id_3")
```

Comment:
Maybe we should also add a test with a WHERE clause (not ON condition). Just to double check if the conditions are properly separated: https://stackoverflow.com/questions/354070/sql-join-what-is-the-difference-between-where-clause-and-on-clause

Reply:
Yeah, we should to test the MJ with filters. I'd have to pass down postJoinFilter and use it in the operator so we push down the filter in this operator in this PR, might be simple. Or else, we can not push it down for now and have it in the calc